ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -1284,56 +1292,60 @@ RouterErrorCode processServerError(ServerErrorCode errorCode) {
      */
     private void handleMetadataBlob(BlobData blobData, byte[] userMetadata, ByteBuffer encryptionKey)
         throws IOException, MessageFormatException {
-      ByteBuffer serializedMetadataContent = blobData.getStream().getByteBuffer();
-      compositeBlobInfo =
-          MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent, blobIdFactory);
-      totalSize = compositeBlobInfo.getTotalSize();
-      chunkMetadataList = compositeBlobInfo.getChunkMetadataList();
-      boolean rangeResolutionFailure = false;
+      ByteBuf serializedMetadataContent = blobData.getAndRelease();
       try {
-        if (options.getBlobOptions.hasBlobSegmentIdx()) {
-          int requestedSegment = options.getBlobOptions.getBlobSegmentIdx();
-          if (requestedSegment < 0 || requestedSegment >= chunkMetadataList.size()) {
-            throw new IllegalArgumentException(
-                "Bad segment number: " + requestedSegment + ", num of keys: " + chunkMetadataList.size());
+        compositeBlobInfo =
+            MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent.nioBuffer(), blobIdFactory);
```

Comment:
Maybe in another PR, `deserializeMetadataContentRecord` can be changed to take in a `ByteBuf`.

Reply:
Totally agree, then we will have to change lots of methods. Planning on doing that in a different PR